Feature/book reading status

Imjurney opened the pull request:
- 1단계 - 도서 기본 정보, 독서 상태, 독서할 시작 및 종료일
- 2단계 - 도서 추천 여부, 별점에 대해 로직을 완료했습니다.

Imjurney committed:
feat: 알라딘 API 검색 기능 추가 및 관련 컴포넌트 및 타입 정의 수정

Imjurney committed:
feat: 독서 상태 기능 추가

Imjurney committed:
feat: RH 컴포넌트추가,step01~02 레이어 완성

virere commented on src/components/BookSearchResult/BookSearchResult.tsx:
```diff
@@ -0,0 +1,126 @@
+  onSelectBook,
+  isSelectedBook,
+}: BookSearchResultProps) => {
+  if (isLoading) return <p css={SearchTextStyle}>로딩 중...</p>;
```
suspense 를 사용하면 어떨까요?

virere commented on src/components/BookSearchResult/BookSearchResult.tsx:
```diff
@@ -0,0 +1,126 @@
+  isSelectedBook,
+}: BookSearchResultProps) => {
+  if (isLoading) return <p css={SearchTextStyle}>로딩 중...</p>;
+  if (books.length > 0)
```
truthy 계산 보다는 books length 가 0일때에 대한 UI를 따로 관리하고 여기는 결과만 보여주는 것이 나을 것 같아요

brightparagon replied:
truthy 계산 보다는 books length 가 0일때에 대한 UI를 따로 관리하고 여기는 결과만 보여주는 것이 나을 것 같아요

brightparagon commented on src/components/BookSearchResult/BookSearchResult.tsx:
```diff
@@ -0,0 +1,126 @@
+  onSelectBook,
+  isSelectedBook,
+}: BookSearchResultProps) => {
+  if (isLoading) return <p css={SearchTextStyle}>로딩 중...</p>;
```
suspense를 쓰면 어떨까요?

Imjurney replied:
tanstack v5부터 suspense기능을 useQuery에서 사용할 수가 없습니다.
useSuspenseQuery를 사용해야하는데, Enabled이 없다보니 초기에 제어를 할 수가 없습니다.

상태에대한 건 따로 관리하되, isLoading을 그대로 써보려하는데 어떨까요?

brightparagon commented on src/components/BookSearchResult/BookSearchResult.tsx:
```diff
@@ -0,0 +1,126 @@
+    return <p css={SearchTextStyle}> 검색 결과가 없습니다.</p>;
+  return null;
+};
+const SearchTextStyle = css`
```
css prop은 재활용되지 않는한 jsx 상에서 바로 쓰는게 가독성에 도움이 됩니다

brightparagon commented on src/components/RHTextarea/RHTextarea.tsx:
```diff
@@ -0,0 +1,91 @@
+          {length} / {maxLength}
+        </span>
+      </div>
+      {error && (
```
분명한 null check가 필요합니다

brightparagon commented on src/components/RHcalendarPicker/RHcalendarPicker.tsx:
```diff
@@ -0,0 +1,73 @@
+              {label}
+            </label>
+          )}
+          <ReactDatePicker
```
date picker를 직접 구현해보시면 어떨까요? 실제 환경에서는 라이브러리를 쓰지 않을거예요

brightparagon replied:
쉽게는 input 만으로 할 수도 있습니다

brightparagon commented on src/components/step/step01.tsx:
```diff
@@ -0,0 +1,253 @@
+  selectPubDate: string | null;
+}
+
+const Step01 = ({
+  isSelectedBook,
+  isLoading,
+  bookList,
+  searchQuery,
+  selectedStatus,
+  handleSearch,
+  handleSelectBook,
+  handleResetBook,
+  handleStatusClick,
+  selectPubDate,
+}: Step01Props) => {
```
컴포넌트가 받아야하는 props가 지나치게 많습니다. props를 받지 않아도 되는 방향으로 재설계를 해볼 수 있을 것 같아요

brightparagon commented on src/hooks/useBookSearch.ts:
```diff
@@ -0,0 +1,38 @@
+  return useQuery({
+    queryKey: ['aladinSearch', query],
+    queryFn: async () => {
+      const params = new URLSearchParams({
+        query: query,
+        queryType: 'Keyword',
+        maxResults: '20',
+        start: '1',
+        sort: 'Accuracy',
+        cover: 'Medium',
+      });
```
파라미터는 훅이 직접 읽으면 위험해요. 페이지나 컴포넌트 레벨에서 읽고 훅에 전달하는 방식이어야 안전합니다

brightparagon commented on src/pages/register/index.tsx:
```diff
@@ -0,0 +1,178 @@
+  { ssr: false }
+);
+
+export default function Register() {
```
- 쿼리 파라미터를 다루는 코드
- 라우팅을 다루는 코드
- api 조회 필터를 다루는 코드
- 폼 코드

한 페이지에 모든 것들이 섞여있어요. 관련된 것들끼리 따로 훅이나 컴포넌트로 분리하고 페이지에서 재활용하는 방식으로 리팩토링 해볼 수 있을 것 같아요

brightparagon commented on src/pages/register/index.tsx:
```diff
@@ -0,0 +1,178 @@
+            onSubmit={handleSubmit(data => console.log(data))}
+          >
+            {/*1. 도서 검색 및 기본 선택 영역 */}
+            {step === 1 && (
+              <Step01
+                isSelectedBook={isSelectedBook}
+                isLoading={isLoading}
+                bookList={bookList}
+                searchQuery={searchQuery}
+                selectedStatus={selectedStatus}
+                handleSearch={handleSearch}
+                handleSelectBook={handleSelectBook}
+                handleResetBook={handleResetBook}
+                handleStatusClick={handleStatusClick}
+                selectPubDate={selectPubDate}
+              />
+            )}
+            {/*2. 도서 추천 여부, 별점*/}
+            {step === 2 && <Step02 />}
```
이런 방식으로 선언적으로 관리해보는 방법을 고민해볼 수 있을 것 같아요

```
<SwitchCases
  value={step}
  cases={{
    step1: <Step1 />
    ...
  }}
/>
```

brightparagon commented on src/components/step/step01.tsx:
```diff
@@ -0,0 +1,253 @@
+            required={!isSelectedBook}
+          />
+
+          <Button
+            disabled={isLoading}
+            onClick={handleSearch}
+            addcss={ButtonPosition}
+            type='button'
+            variant='primary'
+            size='sm'
+            iconPosition='left'
+            icon={<IoSearchOutline />}
+          >
+            검색
+          </Button>
```
요구사항에 검색이라던가 선택하는 것은 없었는데 이건 왜 들어가는건가요?

Imjurney replied:
@brightparagon 
네 요구사항에 검색은 없었습니다. 다만 해당 도서의 정확한 정보를 가져오는게 맞다고 생각이 들어서 관련
오픈 API를 찾아 적용하는 과정에서 생겼던 것 같습니다. 입력된 인풋을 실시간으로 가져오는 response 객체가 아닌 것 같아 검색 방식으로 최초로 앞단을 추가했어요. 1번 요구사항에 맞춰서 다시 세팅하겠습니다.

virere replied:
🙆‍♂️

brightparagon replied:
🙆‍♂️

Imjurney committed:
feat: 패키지 의존성 추가 및 불필요한 의존성 제거, RH 컴포넌트 개선

Imjurney committed:
Refactor book selection and feedback steps in registration process

- Simplified BookSearchResult component by removing loading state handling.
- Enhanced RHTextarea to focus on feedback when specific ratings are selected.
- Improved RHCalendarPicker to handle date input more intuitively.
- Updated StarFeedback component to remove unnecessary feedback logic.
- Restructured Step01 to utilize new BookSearch and SelectedBookItem components for better readability.
- Added DateRangePicker component to manage date selection based on book status.
- Implemented form state management using Jotai for persistent form data.
- Created new components for book search and results display.
- Introduced validation schemas for each step of the registration process.
- Added steps for feedback and confirmation in the registration flow.
